lib/subproviders/geth_api_double.js

```diff
@@ -568,30 +568,83 @@ GethApiDouble.prototype.eth_getUncleCountByBlockNumber = function(blockNumber, c
 };
 
 /**
- * Returns information about a uncle of a block by hash and uncle index position.
+ * Returns the number of uncles in a block matching the given block hash.
  *
- * @param {DATA, 32 Bytes} hash - hash of a block
- * @param {QUANTITY} index - the uncle's index position.
+ * @param {string} hash - Hash of a block (32 bytes).
+ * @param {number} index - Requested index to uncle array
  * @callback callback
- * @param {error} err - Error Object
- * @param {Object} result - A block object,
+ * @throws - Block hash not found
+ * @throws {OutOfRange} - Uncle array index must be 0 or 1
+ * @throws {OutOfRange} - Uncle array index is out of range
+ * @returns {number} result - Integer count of uncles in this block.
  */
 GethApiDouble.prototype.eth_getUncleByBlockHashAndIndex = function(hash, index, callback) {
-  callback(null, {});
+  // Check for valid uncle index
+  if (index !== 0 && index !== 1) {
+    const error = new Error("Uncle array index must be 0 or 1");
+    callback(error);
+    return;
+  }
+
+  this.state.getBlock(hash, function(err, block) {
+    if (err) {
+      if (err.message && err.message.indexOf("Key not found in database") >= 0) {
+        const error = new Error("Unknown block hash");
+        callback(error);
+      } else {
+        callback(err);
+      }
+    } else {
+      const { uncleHeaders } = block;
+      const uncleCount = uncleHeaders.length;
+      if (index < uncleCount) {
+        callback(null, uncleHeaders[index]);
+      } else {
+        const error = new Error("Uncle array index is out of range");
+        callback(error);
+      }
+    }
+  });
 };
 
 /**
- * Returns information about a uncle of a block by number and uncle index position.
+ * Returns the number of uncles in a block matching the given block number.
  *
- * @param {QUANTITY} blockNumber -
- * ^a block number, or the string "earliest", "latest" or "pending", as in the default block parameter.
- * @param {QUANTITY} uncleIndex - the uncle's index position.
+ * @param {number|string} blockNumber - Integer block number, "latest", "earliest" or "pending"
  * @callback callback
- * @param {error} err - Error object
- * @param {Object} resutl - A block object,
+ * @throws - Uncle array index must be 0 or 1
+ * @throws {OutOfRange} - Block number is either negative or higher than block height
+ * @throws - Invalid tag detected
+ * @throws {OutOfRange} Uncle array index is out of range
+ * @returns {number} result - Integer count of uncles in this block.
  */
-GethApiDouble.prototype.eth_getUncleByBlockNumberAndIndex = function(blockNumber, uncleIndex, callback) {
-  callback(null, {});
+GethApiDouble.prototype.eth_getUncleByBlockNumberAndIndex = function(blockNumber, index, callback) {
+  // Check for valid uncle index
+  if (index !== 0 && index !== 1) {
+    const error = new Error("Uncle array index must be 0 or 1");
+    callback(error);
+    return;
+  }
+
+  this.state.getBlock(blockNumber, function(err, block) {
```

Comment:
Can flatten these ifs? Im not sure if that would be more or less readable..  Something like:
```
  this.state.getBlock(hash, function(err, block) {
    let error = err;
    if (err && err.message && err.message.indexOf("Key not found in database") >= 0) {
      error = new Error("Unknown block hash");
    } else if (index >= block.uncleHeaders.length) {
      error = new Error("Uncle array index is out of range");
    } else {
      callback(null, uncleHeaders[index]);
      return;
    }
    callback(error);
  });
```

Reply:
That can be updated.

Reply:
On second thought, this may make things more unreadable considering that `callback` accept any parameter(s).

Reply:
My thought process here is eliminating the number of `callback()` calls.  Such that it either succeeds or it doesn't and reduces the amount of nested blocks which (maybe?) makes it easier to follow.  I wrote this quickly and didn't test this or anything and I realize now I may have missed something and in addition maybe this can be simplified even further.  I'll test quickly in the morning with a fresh pair of eyes.

Reply:
So what I originally suggested was slightly flawed - if err exists but the rest of the conditional fails we might not return an error as expected.  Either way - just some ideas..
```
 this.state.getBlock(hash, function(err, block) {
    let error;
    if (err) {
        error = err.message && err.message.indexOf("Key not found in database") >= 0
            ? new Error("Unknown block hash")
            : err;
    } else if (index >= block.uncleHeaders.length) {
      error = new Error("Uncle array index is out of range");
    } else {
      callback(null, uncleHeaders[index]);
    }
    callback(error);
  });
```
or with *1* callback call but likely even less readable:
```
 this.state.getBlock(hash, function(err, block) {
    let params;
    if (err) {
        params = err.message && err.message.indexOf("Key not found in database") >= 0
            ? [new Error("Unknown block hash")]
            : [err];
    } else if (index >= block.uncleHeaders.length) {
      params = [new Error("Uncle array index is out of range")];
    } else {
      params = [null, uncleHeaders[index]];
    }
    callback(...params);
  });
```

Reply:
This may break a linting rule but may be worth the override.

Reply:
If we have to override the linter then the changes aren't worth it in my opinion.  Let's just aim for reducing the number of callback()'s and let's opt for the early return on invalid index's rather than wrapping the block in a conditional.